File: sdk/python/ai_runtime/ast/engine.py
"""
ai_runtime.ast.engine
~~~~~~~~~~~~~~~~~~~~~
AST-based application engine using libcst to preserve formatting and comments.
"""

from __future__ import annotations

import textwrap
from pathlib import Path
import libcst as cst
from libcst.metadata import PositionProvider

def apply_patch(patch: dict, project_root: str) -> None:
    """
    Applies the patch to the target file.
    """
    target = patch.get("target", {})
    action = patch.get("action", "")
    file_path = Path(project_root) / target["file"]
    
    if not file_path.exists():
        if action in ("add_function", "modify_class") and not file_path.parent.exists():
            file_path.parent.mkdir(parents=True, exist_ok=True)
            file_path.touch()
        elif action != "add_function":
            raise FileNotFoundError(f"Target file not found: {file_path}")
            
    # Read original source
    source = file_path.read_text(encoding="utf-8") if file_path.exists() else ""
    if not source.strip():
        # Handle empty files by providing a base structure
        source = ""

    # Parse with libcst
    module = cst.parse_module(source)
    wrapper = cst.MetadataWrapper(module)
    
    # Dispatch
    if action == "modify_function":
        module = _apply_modify_symbol(wrapper, target, patch.get("changes", {}), "function")
    elif action == "add_function":
        module = _apply_add_function(module, target, patch.get("changes", {}))
    elif action == "remove_function":
        module = _apply_remove_symbol(wrapper, target, "function")
    elif action == "modify_class":
        module = _apply_modify_symbol(wrapper, target, patch.get("changes", {}), "class")
    elif action == "update_import":
        module = _apply_update_import(module, patch.get("changes", {}))
    elif action == "replace_block":
        module = _apply_replace_block(module, patch.get("changes", {}))
    else:
        raise ValueError(f"Unsupported action: {action}")
        
    # Write back
    file_path.write_text(module.code, encoding="utf-8")


class SymbolModifier(cst.CSTTransformer):
    METADATA_DEPENDENCIES = (PositionProvider,)

    def __init__(self, symbol_name: str, symbol_type: str, operation: str, payload: str):
        self.symbol_name = symbol_name
        self.symbol_type = symbol_type
        self.operation = operation
        self.payload = textwrap.dedent(payload).strip()
        self.found = False

    def _handle_node(self, original_node, updated_node):
        if original_node.name.value != self.symbol_name:
            return updated_node

        self.found = True
        pos = self.get_metadata(PositionProvider, original_node)
        indent = " " * pos.start.column

        if self.operation in ("replace_body", "update_logic"):
            indented_payload = textwrap.indent(self.payload, indent + "    ")
            # Create a dummy container based on type
            if isinstance(original_node, cst.ClassDef):
                dummy_code = f"class Dummy:\n{indented_payload}"
            else:
                dummy_code = f"def dummy():\n{indented_payload}"
            
            try:
                dummy_module = cst.parse_module(dummy_code)
                new_body = dummy_module.body[0].body
                return updated_node.with_changes(body=new_body)
            except Exception as e:
                raise ValueError(f"Failed to parse new body for {self.symbol_name}: {e}")

        elif self.operation in ("insert_before", "insert_after"):
            indented_payload = textwrap.indent(self.payload, indent)
            try:
                dummy_module = cst.parse_module(indented_payload)
                new_nodes = dummy_module.body
            except Exception as e:
                raise ValueError(f"Failed to parse payload for {self.symbol_name}: {e}")
            
            if self.operation == "insert_before":
                return cst.FlattenSentinel([*new_nodes, updated_node])
            else:
                return cst.FlattenSentinel([updated_node, *new_nodes])

        return updated_node

    def leave_FunctionDef(self, original_node: cst.FunctionDef, updated_node: cst.FunctionDef):
        if self.symbol_type == "function":
            return self._handle_node(original_node, updated_node)
        return updated_node

    def leave_ClassDef(self, original_node: cst.ClassDef, updated_node: cst.ClassDef):
        if self.symbol_type == "class":
            return self._handle_node(original_node, updated_node)
        return updated_node


class SymbolRemover(cst.CSTTransformer):
    def __init__(self, target_name: str, symbol_type: str):
        self.target_name = target_name
        self.symbol_type = symbol_type
        self.removed = False

    def _handle_node(self, original_node, updated_node):
        if original_node.name.value == self.target_name:
            self.removed = True
            return cst.RemoveFromParent()
        return updated_node

    def leave_FunctionDef(self, original_node: cst.FunctionDef, updated_node: cst.FunctionDef):
        if self.symbol_type == "function":
            return self._handle_node(original_node, updated_node)
        return updated_node

    def leave_ClassDef(self, original_node: cst.ClassDef, updated_node: cst.ClassDef):
        if self.symbol_type == "class":
            return self._handle_node(original_node, updated_node)
        return updated_node


class ImportRemover(cst.CSTTransformer):
    def __init__(self, imports_to_remove: list[str]):
        self.target_strings = set(imp.strip() for imp in imports_to_remove)

    def leave_Import(self, original_node: cst.Import, updated_node: cst.Import):
        code = cst.Module(body=[original_node]).code.strip()
        if code in self.target_strings:
            return cst.RemoveFromParent()
        return updated_node

    def leave_ImportFrom(self, original_node: cst.ImportFrom, updated_node: cst.ImportFrom):
        code = cst.Module(body=[original_node]).code.strip()
        if code in self.target_strings:
            return cst.RemoveFromParent()
        return updated_node


def _apply_modify_symbol(wrapper: cst.MetadataWrapper, target: dict, changes: dict, symbol_type: str) -> cst.Module:
    operation = changes.get("operation")
    payload = changes.get("payload", "")
    symbol = target.get("symbol")
    
    if not symbol:
        raise ValueError(f"modify_{symbol_type} requires target.symbol")
        
    transformer = SymbolModifier(symbol, symbol_type, operation, payload)
    new_module = wrapper.visit(transformer)
    if not transformer.found:
        raise ValueError(f"{symbol_type.capitalize()} {symbol} not found in target file")
    return new_module

def _apply_remove_symbol(wrapper: cst.MetadataWrapper, target: dict, symbol_type: str) -> cst.Module:
    symbol = target.get("symbol")
    if not symbol:
        raise ValueError(f"remove_{symbol_type} requires target.symbol")
        
    transformer = SymbolRemover(symbol, symbol_type)
    new_module = wrapper.visit(transformer)
    if not transformer.removed:
        raise ValueError(f"{symbol_type.capitalize()} {symbol} not found to remove")
    return new_module

def _apply_add_function(module: cst.Module, target: dict, changes: dict) -> cst.Module:
    payload = changes.get("payload", "").strip()
    try:
        new_func_module = cst.parse_module(payload)
        new_func = new_func_module.body[0]
    except Exception as e:
        raise ValueError(f"Failed to parse new function: {e}")
        
    new_body = list(module.body)
    if new_body:
        new_func = new_func.with_changes(leading_lines=[cst.EmptyLine()])
        
    new_body.append(new_func)
    return module.with_changes(body=new_body)

def _apply_update_import(module: cst.Module, changes: dict) -> cst.Module:
    operation = changes.get("operation")
    imports = changes.get("imports", [])
    
    if operation == "add_import":
        parsed_imports = []
        for imp in imports:
            try:
                mod = cst.parse_module(imp)
                parsed_imports.extend(mod.body)
            except Exception as e:
                raise ValueError(f"Failed to parse import '{imp}': {e}")
                
        new_body = list(module.body)
        insert_idx = _import_insert_index(new_body)
                
        new_body = new_body[:insert_idx] + parsed_imports + new_body[insert_idx:]
        return module.with_changes(body=new_body)
    elif operation == "remove_import":
        transformer = ImportRemover(imports)
        return module.visit(transformer)
    else:
        raise ValueError(f"Unsupported operation for update_import: {operation}")


def _import_insert_index(body: list[cst.CSTNode]) -> int:
    insert_idx = 0
    saw_import_block = False

    for i, stmt in enumerate(body):
        first = _first_small_statement(stmt)
        if first is None:
            if insert_idx == 0 and _is_module_docstring(stmt):
                insert_idx = i + 1
                continue
            if saw_import_block:
                break
            continue

        if _is_future_import(first):
            insert_idx = i + 1
            continue

        if isinstance(first, (cst.Import, cst.ImportFrom)):
            insert_idx = i + 1
            saw_import_block = True
            continue

        if saw_import_block:
            break

    return insert_idx


def _first_small_statement(stmt: cst.CSTNode) -> cst.CSTNode | None:
    if not isinstance(stmt, cst.SimpleStatementLine) or not stmt.body:
        return None
    return stmt.body[0]


def _is_future_import(stmt: cst.CSTNode) -> bool:
    return (
        isinstance(stmt, cst.ImportFrom)
        and isinstance(stmt.module, cst.Name)
        and stmt.module.value == "__future__"
    )


def _is_module_docstring(stmt: cst.CSTNode) -> bool:
    first = _first_small_statement(stmt)
    return (
        isinstance(first, cst.Expr)
        and isinstance(first.value, cst.SimpleString)
    )

def _apply_replace_block(module: cst.Module, changes: dict) -> cst.Module:
    operation = changes.get("operation")
    if operation != "context_replace":
        raise ValueError(f"Unsupported operation for replace_block: {operation}")
        
    context_before = changes.get("context_before", "")
    context_after = changes.get("context_after", "")
    payload = changes.get("payload", "")
    
    if not context_before:
        raise ValueError("replace_block requires context_before")
        
    source = module.code
    before_idx = source.find(context_before)
    if before_idx == -1:
        raise ValueError("context_before not found in source")
        
    if context_after:
        after_idx = source.find(context_after, before_idx + len(context_before))
        if after_idx == -1:
            raise ValueError("context_after not found in source")
        new_source = source[:before_idx + len(context_before)] + "\n" + payload + "\n" + source[after_idx:]
    else:
        new_source = source[:before_idx + len(context_before)] + "\n" + payload
        
    try:
        return cst.parse_module(new_source)
    except Exception as e:
        raise ValueError(f"Failed to parse file after replace_block: {e}")

File: sdk/node/src/ast/engine.ts
import * as fs from 'fs';
import * as path from 'path';
import * as recast from 'recast';
import * as tsParser from 'recast/parsers/typescript';

function getParser() {
    return tsParser;
}

function parseReplacementBody(payload: string): any {
    const dummyCode = `function dummy() {\n${payload}\n}`;
    try {
        const dummyAst = recast.parse(dummyCode, { parser: getParser() });
        return dummyAst.program.body[0].body;
    } catch (functionError: any) {
        const privateNames = Array.from(
            new Set(Array.from(payload.matchAll(/this\.#([A-Za-z_$][\w$]*)/g), match => match[1])),
        );
        if (privateNames.length === 0) {
            throw functionError;
        }

        const declarations = privateNames.map(name => `#${name};`).join("\n");
        const dummyClassCode = `class Dummy {\n${declarations}\ndummy() {\n${payload}\n}\n}`;
        const dummyAst = recast.parse(dummyClassCode, { parser: getParser() });
        const classBody = dummyAst.program.body[0].body.body;
        const dummyMethod = classBody.find((node: any) => node.key?.name === "dummy");
        if (!dummyMethod) {
            throw new Error("Could not extract dummy class method body");
        }
        return dummyMethod.body;
    }
}

export function applyPatch(patch: any, projectRoot: string): void {
    const target = patch.target || {};
    const action = patch.action || "";
    const filePath = path.join(projectRoot, target.file);
    
    if (!fs.existsSync(filePath)) {
        if ((action === "add_function" || action === "modify_class") && !fs.existsSync(path.dirname(filePath))) {
            fs.mkdirSync(path.dirname(filePath), { recursive: true });
            fs.writeFileSync(filePath, "", "utf-8");
        } else if (action !== "add_function") {
            throw new Error(`Target file not found: ${filePath}`);
        }
    }
    
    let source = "";
    if (fs.existsSync(filePath)) {
        source = fs.readFileSync(filePath, "utf-8");
    }
    
    if (!source.trim()) {
        source = "";
    }
    
    let ast: any;
    try {
        ast = recast.parse(source, { parser: getParser() });
    } catch (e: any) {
        throw new Error(`Failed to parse file: ${e.message}`);
    }
    
    const changes = patch.changes || {};
    
    if (action === "modify_function") {
        applyModifyFunction(ast, target, changes);
    } else if (action === "add_function") {
        applyAddFunction(ast, target, changes);
    } else if (action === "remove_function") {
        applyRemoveFunction(ast, target);
    } else if (action === "modify_class") {
        applyModifyClass(ast, target, changes);
    } else if (action === "update_import") {
        applyUpdateImport(ast, changes);
    } else if (action === "replace_block") {
        ast = applyReplaceBlock(ast, changes, source);
    } else {
        throw new Error(`Unsupported action: ${action}`);
    }
    
    const output = recast.print(ast).code;
    fs.writeFileSync(filePath, output, "utf-8");
}

function applyModifyFunction(ast: any, target: any, changes: any): void {
    const operation = changes.operation;
    const payload = changes.payload || "";
    const symbol = target.symbol;
    
    if (!symbol) {
        throw new Error("modify_function requires target.symbol");
    }
    
    if (operation === "replace_body" || operation === "update_logic") {
        let newBody: any;
        try {
            newBody = parseReplacementBody(payload);
        } catch (e: any) {
            throw new Error(`Failed to parse new function body: ${e.message}`);
        }
        
        let found = false;
        recast.visit(ast, {
            visitFunctionDeclaration(path) {
                if (path.node.id?.type === 'Identifier' && path.node.id.name === symbol) {
                    found = true;
                    path.node.body = newBody;
                    return false;
                }
                this.traverse(path);
            },
            visitFunctionExpression(path) {
                if (path.node.id?.type === 'Identifier' && path.node.id.name === symbol) {
                    found = true;
                    path.node.body = newBody;
                    return false;
                }
                this.traverse(path);
            },
            visitArrowFunctionExpression(path) {
                if (path.parentPath?.node?.type === 'VariableDeclarator' && path.parentPath.node.id?.type === 'Identifier' && path.parentPath.node.id.name === symbol) {
                    found = true;
                    path.node.body = newBody;
                    return false;
                }
                this.traverse(path);
            },
            visitClassMethod(path) {
                if (path.node.key?.type === 'Identifier' && path.node.key.name === symbol) {
                    found = true;
                    path.node.body = newBody;
                    return false;
                }
                this.traverse(path);
            }
        });
        
        if (!found) {
            throw new Error(`Function ${symbol} not found in target file`);
        }
    } else if (operation === "insert_before" || operation === "insert_after") {
        let newNodes: any[];
        try {
            const parsed = recast.parse(payload, { parser: getParser() });
            newNodes = parsed.program.body;
        } catch (e: any) {
            throw new Error(`Failed to parse payload for insert: ${e.message}`);
        }
        
        let found = false;
        recast.visit(ast, {
            visitFunctionDeclaration(path) {
                if (path.node.id?.type === 'Identifier' && path.node.id.name === symbol) {
                    found = true;
                    if (operation === "insert_before") {
                        path.insertBefore(...newNodes);
                    } else {
                        path.insertAfter(...newNodes);
                    }
                    return false;
                }
                this.traverse(path);
            },
            visitFunctionExpression(path) {
                if (path.node.id?.type === 'Identifier' && path.node.id.name === symbol) {
                    found = true;
                    let targetPath: any = path;
                    while (targetPath && targetPath.node.type !== 'VariableDeclaration') {
                        targetPath = targetPath.parentPath;
                    }
                    if (targetPath && targetPath.node.type === 'VariableDeclaration') {
                        if (operation === "insert_before") {
                            targetPath.insertBefore(...newNodes);
                        } else {
                            targetPath.insertAfter(...newNodes);
                        }
                    } else {
                        throw new Error("Cannot insert_before/after a FunctionExpression without a parent variable declaration");
                    }
                    return false;
                }
                this.traverse(path);
            },
            visitArrowFunctionExpression(path) {
                if (path.parentPath?.node?.type === 'VariableDeclarator' && path.parentPath.node.id?.type === 'Identifier' && path.parentPath.node.id.name === symbol) {
                    found = true;
                    let targetPath: any = path;
                    while (targetPath && targetPath.node.type !== 'VariableDeclaration') {
                        targetPath = targetPath.parentPath;
                    }
                    if (targetPath && targetPath.node.type === 'VariableDeclaration') {
                        if (operation === "insert_before") {
                            targetPath.insertBefore(...newNodes);
                        } else {
                            targetPath.insertAfter(...newNodes);
                        }
                    } else {
                        throw new Error("Cannot insert_before/after an ArrowFunction without a parent variable declaration");
                    }
                    return false;
                }
                this.traverse(path);
            },
            visitClassMethod(path) {
                if (path.node.key?.type === 'Identifier' && path.node.key.name === symbol) {
                    found = true;
                    if (operation === "insert_before") {
                        path.insertBefore(...newNodes);
                    } else {
                        path.insertAfter(...newNodes);
                    }
                    return false;
                }
                this.traverse(path);
            }
        });
        
        if (!found) {
            throw new Error(`Function ${symbol} not found in target file`);
        }
    } else {
        throw new Error(`Operation ${operation} not fully implemented for modify_function`);
    }
}

function applyAddFunction(ast: any, target: any, changes: any): void {
    const payload = (changes.payload || "").trim();
    let newFuncAst: any;
    try {
        newFuncAst = recast.parse(payload, { parser: getParser() });
    } catch (e: any) {
        throw new Error(`Failed to parse new function: ${e.message}`);
    }
    
    const newFuncNodes = newFuncAst.program.body;
    if (!ast.program.body) {
        ast.program.body = [];
    }
    ast.program.body.push(...newFuncNodes);
}

function applyRemoveFunction(ast: any, target: any): void {
    const symbol = target.symbol;
    if (!symbol) {
        throw new Error("remove_function requires target.symbol");
    }
    
    let removed = false;
    recast.visit(ast, {
        visitFunctionDeclaration(path) {
            if (path.node.id?.type === 'Identifier' && path.node.id.name === symbol) {
                removed = true;
                path.prune();
                return false;
            }
            this.traverse(path);
        },
        visitVariableDeclaration(path) {
            let declsToRemove = [];
            for (let i = 0; i < path.node.declarations.length; i++) {
                const decl: any = path.node.declarations[i];
                if (decl.id?.type === 'Identifier' && decl.id.name === symbol && 
                    (decl.init?.type === 'ArrowFunctionExpression' || decl.init?.type === 'FunctionExpression')) {
                    removed = true;
                    declsToRemove.push(i);
                }
            }
            if (declsToRemove.length > 0) {
                if (declsToRemove.length === path.node.declarations.length) {
                    path.prune();
                } else {
                    // Remove the specific declarators
                    // Reverse to keep indices valid
                    for (const idx of declsToRemove.reverse()) {
                        path.node.declarations.splice(idx, 1);
                    }
                }
                return false;
            }
            this.traverse(path);
        },
        visitClassMethod(path) {
            if (path.node.key?.type === 'Identifier' && path.node.key.name === symbol) {
                removed = true;
                path.prune();
                return false;
            }
            this.traverse(path);
        }
    });
    
    if (!removed) {
        throw new Error(`Function ${symbol} not found to remove`);
    }
}

function applyUpdateImport(ast: any, changes: any): void {
    const operation = changes.operation;
    const imports = changes.imports || [];
    
    if (operation === "add_import") {
        const parsedImports: any[] = [];
        for (const imp of imports) {
            try {
                const mod = recast.parse(imp, { parser: getParser() });
                parsedImports.push(...mod.program.body);
            } catch (e: any) {
                throw new Error(`Failed to parse import '${imp}': ${e.message}`);
            }
        }
        
        let insertIdx = 0;
        const body = ast.program.body || [];
        for (let i = 0; i < body.length; i++) {
            if (body[i].type === 'ImportDeclaration') {
                insertIdx = i + 1;
            }
        }
        
        body.splice(insertIdx, 0, ...parsedImports);
    } else if (operation === "remove_import") {
        const targetStrings = new Set(imports.map((i: string) => i.trim()));
        recast.visit(ast, {
            visitImportDeclaration(path) {
                const code = recast.print(path.node).code.trim();
                // Simple matching or check if any of the target strings matches
                if (targetStrings.has(code) || targetStrings.has(code + ';')) {
                    path.prune();
                    return false;
                }
                this.traverse(path);
            }
        });
    } else {
        throw new Error(`Unsupported operation for update_import: ${operation}`);
    }
}

function applyModifyClass(ast: any, target: any, changes: any): void {
    const symbol = target.symbol;
    const operation = changes.operation;
    const payload = changes.payload || "";

    if (!symbol) {
        throw new Error("modify_class requires target.symbol");
    }

    if (operation === "replace_body" || operation === "update_logic") {
        const dummyCode = `class Dummy {\n${payload}\n}`;
        let newBody: any;
        try {
            const dummyAst = recast.parse(dummyCode, { parser: getParser() });
            newBody = dummyAst.program.body[0].body;
        } catch (e: any) {
            throw new Error(`Failed to parse new class body: ${e.message}`);
        }

        let found = false;
        recast.visit(ast, {
            visitClassDeclaration(path) {
                if (path.node.id?.name === symbol) {
                    found = true;
                    path.node.body = newBody;
                    return false;
                }
                this.traverse(path);
            },
            visitClassExpression(path) {
                if (path.node.id?.name === symbol) {
                    found = true;
                    path.node.body = newBody;
                    return false;
                }
                this.traverse(path);
            }
        });

        if (!found) {
            throw new Error(`Class ${symbol} not found in target file`);
        }
    } else if (operation === "insert_before" || operation === "insert_after") {
        let newNodes: any[];
        try {
            const parsed = recast.parse(payload, { parser: getParser() });
            newNodes = parsed.program.body;
        } catch (e: any) {
            throw new Error(`Failed to parse payload for insert: ${e.message}`);
        }
        
        let found = false;
        recast.visit(ast, {
            visitClassDeclaration(path) {
                if (path.node.id?.name === symbol) {
                    found = true;
                    if (operation === "insert_before") {
                        path.insertBefore(...newNodes);
                    } else {
                        path.insertAfter(...newNodes);
                    }
                    return false;
                }
                this.traverse(path);
            },
            visitClassExpression(path) {
                if (path.node.id?.name === symbol) {
                    found = true;
                    let targetPath: any = path;
                    while (targetPath && targetPath.node.type !== 'VariableDeclaration') {
                        targetPath = targetPath.parentPath;
                    }
                    if (targetPath && targetPath.node.type === 'VariableDeclaration') {
                        if (operation === "insert_before") {
                            targetPath.insertBefore(...newNodes);
                        } else {
                            targetPath.insertAfter(...newNodes);
                        }
                    } else {
                        throw new Error("Cannot insert_before/after a ClassExpression without a parent variable declaration");
                    }
                    return false;
                }
                this.traverse(path);
            }
        });
        
        if (!found) {
            throw new Error(`Class ${symbol} not found in target file`);
        }
    } else {
        throw new Error(`Operation ${operation} not fully implemented for modify_class`);
    }
}

function applyReplaceBlock(ast: any, changes: any, source: string): any {
    const contextBefore = changes.context_before || "";
    const contextAfter = changes.context_after || "";
    const payload = changes.payload || "";

    if (!contextBefore) throw new Error("replace_block requires context_before");

    const beforeIdx = source.indexOf(contextBefore);
    if (beforeIdx === -1) throw new Error("context_before not found in source");

    let newSource = "";
    if (contextAfter) {
        const afterIdx = source.indexOf(contextAfter, beforeIdx + contextBefore.length);
        if (afterIdx === -1) throw new Error("context_after not found in source");
        newSource = source.substring(0, beforeIdx + contextBefore.length) + "\n" + payload + "\n" + source.substring(afterIdx);
    } else {
        newSource = source.substring(0, beforeIdx + contextBefore.length) + "\n" + payload;
    }

    try {
        return recast.parse(newSource, { parser: getParser() });
    } catch (e: any) {
        throw new Error(`Failed to parse file after replace_block: ${e.message}`);
    }
}
